internal/credential/repository_store.go

```diff
@@ -0,0 +1,190 @@
+	switch {
+	case len(projectIds) == 0:
+		return nil, time.Time{}, errors.New(ctx, errors.InvalidParameter, op, "missing project ids")
+	case limit < 1:
+		return nil, time.Time{}, errors.New(ctx, errors.InvalidParameter, op, "missing limit")
+	}
+
+	var inClauses []string
+	var args []any
+	for i, projectId := range projectIds {
+		arg := "project_id_" + strconv.Itoa(i)
+		inClauses = append(inClauses, "@"+arg)
+		args = append(args, sql.Named(arg, projectId))
+	}
+	inClause := strings.Join(inClauses, ", ")
+	whereClause := "project_id in (" + inClause + ")"
+
+	query := fmt.Sprintf(listStoresTemplate, whereClause, limit)
+	if afterItem != nil {
+		query = fmt.Sprintf(listStoresPageTemplate, whereClause, limit)
+		args = append(args,
+			sql.Named("last_item_create_time", afterItem.GetCreateTime()),
+			sql.Named("last_item_id", afterItem.GetPublicId()),
+		)
+	}
+
+	return s.queryStores(ctx, query, args)
+}
+
+// ListRefresh lists credential stores across all subtypes.
+func (s *StoreRepository) ListRefresh(ctx context.Context, projectIds []string, updatedAfter time.Time, afterItem pagination.Item, limit int) ([]Store, time.Time, error) {
+	const op = "credential.(*StoreRepository).list"
+	switch {
+	case len(projectIds) == 0:
+		return nil, time.Time{}, errors.New(ctx, errors.InvalidParameter, op, "missing project ids")
+	case updatedAfter.IsZero():
+		return nil, time.Time{}, errors.New(ctx, errors.InvalidParameter, op, "missing updated after time")
+	case limit < 1:
+		return nil, time.Time{}, errors.New(ctx, errors.InvalidParameter, op, "missing limit")
+	}
+
+	var inClauses []string
+	var args []any
+	for i, projectId := range projectIds {
+		arg := "project_id_" + strconv.Itoa(i)
+		inClauses = append(inClauses, "@"+arg)
+		args = append(args, sql.Named(arg, projectId))
+	}
+	inClause := strings.Join(inClauses, ", ")
+	whereClause := "project_id in (" + inClause + ")"
+
+	query := fmt.Sprintf(listStoresRefreshTemplate, whereClause, limit)
+	args = append(args,
+		sql.Named("updated_after_time", timestamp.New(updatedAfter)),
+	)
+	if afterItem != nil {
+		query = fmt.Sprintf(listStoresRefreshPageTemplate, whereClause, limit)
+		args = append(args,
+			sql.Named("last_item_update_time", afterItem.GetUpdateTime()),
+			sql.Named("last_item_id", afterItem.GetPublicId()),
+		)
+	}
+
+	return s.queryStores(ctx, query, args)
+}
+
+// EstimatedCount estimates the total number of credential stores.
+func (s *StoreRepository) EstimatedCount(ctx context.Context) (int, error) {
+	const op = "credential.(*StoreRepository).EstimatedCount"
+	rows, err := s.writer.Query(ctx, estimateCountStores, nil)
+	if err != nil {
+		return 0, errors.Wrap(ctx, err, op, errors.WithMsg("failed to query total credential stores"))
+	}
+	var count int
+	for rows.Next() {
+		if err := s.writer.ScanRows(ctx, rows, &count); err != nil {
+			return 0, errors.Wrap(ctx, err, op, errors.WithMsg("failed to query total credential stores"))
+		}
+	}
+	return count, nil
+}
```

Comment:
This could also be per-subtype but TBH I didn't think it was worth having this done per subtype-table when this package has to know about both the subtype tables for the list query anyway. This is also easier to do atomically in this package (no need to pass around transactions). Thoughts appreciated.

Reply:
Yea I think doing this here is the better option.